android-pdf-viewer/src/main/java/com/github/barteksc/pdfviewer/CacheManager.java

```diff
@@ -81,18 +82,18 @@ private void makeAFreeSpace() {
     public void cacheThumbnail(PagePart part) {
         synchronized (thumbnails) {
             // If cache too big, remove and recycle
-            if (thumbnails.size() >= THUMBNAILS_CACHE_SIZE) {
+            while (thumbnails.size() >= THUMBNAILS_CACHE_SIZE) {
                 thumbnails.remove(0).getRenderedBitmap().recycle();
             }
 
             // Then add thumbnail
-            thumbnails.add(part);
+            addWithoutDuplicates(thumbnails, part);

```

Comment:
> cant you check before adding to cache

Is [ this ](https://github.com/VERO-Digital-Solutions/AndroidPdfViewerV2/pull/2/commits/9213dcf113d91c5bc8b34cec1b05d2f3484fbadd) what you meant?

Reply:
Not exactly, this snippet will cache a thumbnail as `Part` in case we have it already in our cache due to if check going `TRUE && FALSE ==> FALSE`

```kotlin
 if (part.isThumbnail() && !cacheManager.getPageParts().contains(part)) {
            cacheManager.cacheThumbnail(part);
        } else {
            cacheManager.cachePart(part);
```

Also why are we checking for pageParts and not thumbnail parts?